### PR TITLE
Add support for filtering attribute keys for streams via an exclude list

### PR DESCRIPTION
**Pull request by codeboten**
## Changes

This change adds details around supporting the creation of a stream with a set of excluded attribute keys. This functionality is supported in the [Go implementation](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/metric/instrument.go#L146) via the use of filters. The proposal in this PR is to address an issue in the configuration working group: https://github.com/open-telemetry/opentelemetry-configuration/issues/112

Please provide a brief description of the changes here.

For non-trivial changes, follow the [change proposal process](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CONTRIBUTING.md#proposing-a-change).

* [x] Links to the prototypes (when adding or changing features) 
* [x] [`CHANGELOG.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md) file updated for non-trivial changes
* [x] [`spec-compliance-matrix.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md) updated if necessary

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,9 @@ release.
 
 ### Metrics
 
+- Add support for filtering attribute keys for streams via an exclude list.
+  ([]())
```
Comment (codeboten) [inline CHANGELOG.md:20]:
```suggestion
  ([#4188](https://github.com/open-telemetry/opentelemetry-specification/pull/4188))
```

**Review by jack-berg — APPROVED** [review]

File: specification/metrics/sdk.md [hunk 2]
```diff
@@ -364,6 +364,10 @@ The SDK MUST accept the following stream configuration parameters:
   keys that identify the attributes that MUST be kept, and all other attributes
   MUST be ignored.
 
+  Additionally, implementations SHOULD support configuring an exclude-list of
```
Comment (arminru) [inline specification/metrics/sdk.md:367]: Please add an entry to https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md for this new SHOULD requirement.

Comment (codeboten) [reply specification/metrics/sdk.md:367]: Added, please review

Comment (codeboten) [reply specification/metrics/sdk.md:367]: @arminru anything else needed here?

**Review by lmolkova — APPROVED** [review]

File: specification/metrics/sdk.md (hunk 2 above)
Comment (cijothomas) [inline specification/metrics/sdk.md:367]: the placement of this looks a bit odd. Perhaps it should be moved after the `attribute_keys` section. i.e line 381. Else this is intermixed with the allow attribute_keys.

Comment (cijothomas) [reply specification/metrics/sdk.md:367]: wondering if we should clarify conflict resolution, when both are specified or if we should recommend to fail-fast instead?
For example, if 'attribute_keys' = ["user_id"] and 'exclude-keys' = ["user_id"] are both existing? Does this call for a fail-fast? OR one wins over other?
The attribute_keys section has a normative wording (_The allow-list contains attribute
  keys that identify the attributes that **MUST** be kept_) about this..

Comment (codeboten) [reply specification/metrics/sdk.md:367]: I've moved the paragraph down and added a line about failing fast, i think that's the best way to resolve such a conflict. PTAL

File: specification/metrics/sdk.md [hunk 3]
```diff
@@ -379,6 +375,12 @@ The SDK MUST accept the following stream configuration parameters:
   parameter configured on the instrument instead. If the `Attributes`
   advisory parameter is absent, all attributes MUST be kept.
 
+  Additionally, implementations SHOULD support configuring an exclude-list of
+  attribute keys. The exclude-list contains attribute keys that identify the
+  attributes that MUST be excluded, all other attributes MUST be kept. If an
+  attribute key is both included and excluded, implementations MAY cause the
```
Comment (cijothomas) [inline specification/metrics/sdk.md:381]: implementations MAY cause the --> the SDK MAY fail fast in accordance with initialization [error handling
  principles](../error-handling.md#basic-error-handling-principles)

Comment (codeboten) [reply specification/metrics/sdk.md]: Updated

**Review by cijothomas — APPROVED** [review]
LGTM, but one note before merging:
I'll let maintainers comment if there is a need to mark this section as experimental, and wait for 2-3 implementations before marking as stable. This PR right now makes no explicit call out, so this is by default treated as stable spec, thought just 1 language has it implemented.

**Review by pellared — APPROVED** [review]

**Review by trask — APPROVED** [review]

**Comment by jack-berg** [comment]
> I'll let maintainers comment if there is a need to mark this section as experimental, and wait for 2-3 implementations before marking as stable. This PR right now makes no explicit call out, so this is by default treated as stable spec, though just 1 language has it implemented.

Its already possible to do this in the stable java implementation. I think the same is true for go. I think this is safe / benign enough to merge as stable without an experimental period.

**Comment by codeboten** [comment]
Don't think the failure is related to this pr:

```
  ERROR: 1 dead links found in ./specification/compatibility/prometheus_and_openmetrics.md !
  [✖] https://github.com/prometheus/prometheus/commit/d9d51c565c622cdc7d626d3e7569652bc28abe15#diff-bdaf80ebc5fa26365f45db53435b960ce623ea6f86747fb8870ad1abc355f64fR76-R83 → Status: 406
```

**Comment by trask** [comment]
> ERROR: 1 dead links found in ./specification/compatibility/prometheus_and_openmetrics.md !
> [✖] https://github.com/prometheus/prometheus/commit/d9d51c565c622cdc7d626d3e7569652bc28abe15#diff-bdaf80ebc5fa26365f45db53435b960ce623ea6f86747fb8870ad1abc355f64fR76-R83 → Status: 406

(re-run resolved for now)

**Review by jsuereth — APPROVED** [review]

**Review by tsloughter — APPROVED** [review]

**Comment by jack-berg** [comment]
@cijothomas since you approved this and you're involved in dotnet / rust, is it safe to say that you don't see this as a problem for dotnet / rust?

@open-telemetry/javascript-maintainers / @open-telemetry/python-maintainers can you comment on whether or not it would be problematic to extend view stream configuration as described in this PR?

**Comment by cijothomas** [comment]
> @cijothomas since you approved this and you're involved in dotnet / rust, is it safe to say that you don't see this as a problem for dotnet / rust?

Correct.

**Review by JamieDanielson — APPROVED** [review]
This seems reasonable. In JS we already have an allow-list, so it would just need to be extended for a deny-list.

**Comment by pichlermarc** [comment]
> This seems reasonable. In JS we already have an allow-list, so it would just need to be extended for a deny-list.

It's already implemented on the next branch and will ship with SDK 2.0 :slightly_smiling_face:

**Pull request merged**